### PR TITLE
Utf8JsonReader: Do not allow non-standard line endings in single-line comments

### PR DESCRIPTION
**Pull request by krwq**
Fixes: https://github.com/dotnet/corefx/issues/33321
Fixes: https://github.com/dotnet/corefx/issues/37931
(also some other fixes - see below)

Some Unicode characters (i.e. U+2028, U+2029) are considered line endings and are sometimes used in some attack vectors when coming over the wire.

This change causes that parser will throw an exception if such character is encountered in a comment.

Also fixes couple of issues w.r.t to single-line comment handling with both mult- and single segment case around EOF/final block which I haven't reported. I.e.:
- end of json on `\r`
- end of json on empty comment in the end (i.e. `{}//` would cause parser to throw)

Perf results are here: dotnet/performance#543

### CODE REVIEW

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 1]
```diff
@@ -2384,21 +2384,37 @@ private bool ConsumeCommentMultiSegment()
...
-                }
-                if (!GetNextSpan())
+                if (IsLastSpan || !GetNextSpan())
                 {
                     if (IsLastSpan)
                     {
-                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.UnexpectedEndOfDataWhileReadingComment);
+                        if (allowEmptyComment)
+                        {
+                            if (_tokenType != JsonTokenType.Comment)
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: The nesting is getting quite deep here. We should invert the condition and break this into a method.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 2]
```diff
@@ -2384,21 +2384,37 @@ private bool ConsumeCommentMultiSegment()
...
+
             localBuffer = localBuffer.Slice(1);
             if (localBuffer.Length == 0)
             {
-                if (IsLastSpan)
-                {
-                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.UnexpectedEndOfDataWhileReadingComment);
-                }
-                if (!GetNextSpan())
+                if (IsLastSpan || !GetNextSpan())
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: This condition flow is hard to follow now and has a duplicate check (when `IsLastSpan` == true).

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs (hunk 1 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: How about:
```C#
if (localBuffer.Length == 0)
{
    if (IsLastSpan)
    {
        ConsumeEmptySingleLineComment(marker);
        return true;
    }

    if (!GetNextSpan())
    {
        if (IsLastSpan)
        {
            ConsumeEmptySingleLineComment(marker);
            return true;
        }
        return false;
    }

    _totalConsumed++;
    _bytePositionInLine++;
    localBuffer = _buffer;
    leftOver = 0;
}
```

```C#
private void ConsumeEmptySingleLineComment(byte marker)
{
    if (marker != JsonConstants.Slash)
    {
        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.UnexpectedEndOfDataWhileReadingComment);
    }

    if (_tokenType != JsonTokenType.Comment)
    {
        _previousTokenType = _tokenType;
    }
    _tokenType = JsonTokenType.Comment;
    _totalConsumed++;
    _bytePositionInLine++;
    _consumed += 2;
}
```

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 3]
```diff
@@ -2384,21 +2384,37 @@ private bool ConsumeCommentMultiSegment()
...
+                        {
+                            if (_tokenType != JsonTokenType.Comment)
+                            {
+                                _previousTokenType = _tokenType;
+                            }
+
+                            _totalConsumed++;
+                            _bytePositionInLine++;
+                            _tokenType = JsonTokenType.Comment;
+                            _consumed += 2;
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: Hmm... I don't think this is correct.

If you check `BytesConsumed`, this ends up setting it incorrectly:
```C#
[Theory]
[InlineData("//", "", 1, 2)] // expected - 2 / actual - 4
[InlineData("//", "", 2, 2)] // expected - 2 / actual - 3
[InlineData("//a", "a", 2, 3)] // expected - 3 / actual - 2
[InlineData("//abc", "abc", 2, 5)] // expected - 5 / actual - 4
public static void JsonWithSingleLineCommentWithNoLineEndingsFinalBlockMultiSegment(string jsonString, string expectedComment, int segmentSize, int consumed)
{
    byte[] dataUtf8 = Encoding.UTF8.GetBytes(jsonString);
    var state = new JsonReaderState(options: new JsonReaderOptions { CommentHandling = JsonCommentHandling.Allow });
    ReadOnlySequence<byte> sequence = JsonTestHelper.GetSequence(dataUtf8, segmentSize);
    var json = new Utf8JsonReader(sequence, isFinalBlock: true, state);

    Assert.True(json.Read());
    Assert.Equal(JsonTokenType.Comment, json.TokenType);
    Assert.Equal(expectedComment, json.GetComment());
    Assert.Equal(consumed, json.BytesConsumed); // this fails for the 4 test cases listed
    Assert.False(json.Read());
}
```

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 4]
```diff
@@ -2515,6 +2563,98 @@ private bool ConsumeSingleLineCommentMultiSegment(ReadOnlySpan<byte> localBuffer
...
+                {
+                    // this can only happen in the end of stream
+                    Debug.Assert(localBuffer.Length < 2);
+                    return -1;
+                }
+            }
+        }
+
+        // assumes first byte (JsonConstants.MaybeDangerousLineSeparator) is already read
+        private void ThrowOnDangerousLineSeparatorMultiSegment(ReadOnlySpan<byte> localBuffer, ref int dangerousLineSeparatorBytesConsumed)
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: This seems unnecessarily complicated, especially compared to the single-segment case. Why do we have a loop here and why do we need to keep passing the int by ref?

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 5]
```diff
@@ -2515,6 +2563,98 @@ private bool ConsumeSingleLineCommentMultiSegment(ReadOnlySpan<byte> localBuffer
...
+                localBuffer = localBuffer.Slice(p);
+                totalIdx += p;
+
+                // or dangerousLineSeparatorBytesConsumed = 1
+                dangerousLineSeparatorBytesConsumed++;
+                ThrowOnDangerousLineSeparatorMultiSegment(localBuffer, ref dangerousLineSeparatorBytesConsumed);
+
+                if (dangerousLineSeparatorBytesConsumed != 0)
+                {
+                    // this can only happen in the end of stream
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: Comment mismatch. We don't have stream.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs (hunk 2 above)
Comment (krwq) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: it's because GetNextSpan changes the state - if IsLastSpan is already true it should be very cheap to check and compiler is possibly able to detect than no writes were done and optimize this path (if not now then perhaps in the future)

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs (hunk 3 above)
Comment (krwq) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: will take a look - thanks!

File: src/System.Text.Json/src/System/Text/Json/JsonConstants.cs [hunk 6]
```diff
@@ -29,6 +29,10 @@ internal static class JsonConstants
         public const byte UtcOffsetToken = (byte)'Z';
         public const byte TimePrefix = (byte)'T';
 
+        // \u2028 and \u2029 are considered respectively line and paragraph separators
+        // UTF-8 representation for them is E2, 80, A8/A9
+        public const byte MaybeDangerousLineSeparator = 0xE2;
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/System.Text.Json/src/System/Text/Json/JsonConstants.cs]: I like the direction, what about `StartingByteOfNonStandardLineSeparator`?

[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/JsonConstants.cs]: Can we drop `Line`? This applies to both line and paragraph separator, so `StartingByteOfNonStandardSeparator`.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: I understand the reasoning, but I would prefer to keep the pattern consistent with the rest of the file, especially given the complexity here.

Elsewhere, we have:

```C#
if(IsLastSpan)
{
   // ...
}

if (!GetNextSpan())
{
   if(IsLastSpan)
   {
      // ...
   }
}
```

**Comment by krwq** [comment]
@ahsonkhan I've rewrote SkipComment logic for multi-segment almost completely and have unified Consume/Skip. Haven't did anything similar for single-segment yet.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs (hunk 5 above)
Comment (krwq) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: missed this comment. Now fixed in my pending changes

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs (hunk 4 above)
Comment (krwq) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: we need to keep track of how many bytes we consumed in the last segment and I didn't want to put GetNextSpan as it would make it even more complicated. I have removed the loop which hopefully made it a bit more legible

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs [hunk 7]
```diff
@@ -2273,6 +2287,43 @@ private bool SkipSingleLineComment(ReadOnlySpan<byte> localBuffer, out int idx)
...
         }
 
+        private int FindLineSeparator(ReadOnlySpan<byte> localBuffer)
+        {
+            int totalIdx = 0;
+            while (true)
+            {
+                int idx = localBuffer.IndexOfAny(JsonConstants.LineFeed, JsonConstants.CarriageReturn, JsonConstants.MaybeDangerousLineSeparator);
+
+                if (idx == -1)
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs]: missed this, pending in my local changes now

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs [hunk 8]
```diff
@@ -2273,6 +2287,43 @@ private bool SkipSingleLineComment(ReadOnlySpan<byte> localBuffer, out int idx)
...
+                    return -1;
+
+                if (localBuffer[idx] != JsonConstants.MaybeDangerousLineSeparator)
+                    return totalIdx + idx;
+
+                int p = idx + 1;
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs]: pending now

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs [hunk 9]
```diff
@@ -2273,6 +2287,43 @@ private bool SkipSingleLineComment(ReadOnlySpan<byte> localBuffer, out int idx)
...
+            // \u2028 and \u2029 are considered respectively line and paragraph separators
+            // UTF-8 representation for them is E2, 80, A8/A9
+            // we have already read E2, we need to check for remaining 2 bytes
+
+            if (localBuffer.Length < 2)
+            {
+                return;
+            }
+
+            if (localBuffer[0] == 0x80 && (localBuffer[1] == 0xA8 || localBuffer[1] == 0xA9))
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs]: pending now

File: src/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs [hunk 10]
```diff
@@ -884,6 +885,139 @@ public static void SkipSingleLineCommentMultiSpanTest(string lineSeparator)
...
+        [InlineData("//a", "a", 2)]
+        [InlineData("//a", "a", 3)]
+        [InlineData("//a", "a", 100)]
+        [InlineData("//abc", "abc", 1)]
+        [InlineData("//abc", "abc", 2)]
+        [InlineData("//abc", "abc", 3)]
+        [InlineData("//abc", "abc", 100)]
+        public static void JsonWithSingleLineCommentWithRegularLineEndingMultiSegment(string jsonStringWithoutLineEnding, string expectedComment, int segmentSize)
+        {
+            foreach (string lineEnding in new string[] { "\r", "\r ", "\r\n", "\r\n ", "\n", "\n " })
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs]: pending now

File: src/System.Text.Json/src/System/Text/Json/JsonConstants.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (krwq) [reply src/System.Text.Json/src/System/Text/Json/JsonConstants.cs]: missed this, pending now

**Comment by ahsonkhan** [comment]
@krwq, reviewing this would take some time and there is a bit of a churn, can this wait till next week? Do you have a sense of perf impact here?

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs [hunk 11]
```diff
@@ -2273,6 +2287,49 @@ private bool SkipSingleLineComment(ReadOnlySpan<byte> localBuffer, out int idx)
...
+            // UTF-8 representation for them is E2, 80, A8/A9
+            // we have already read E2, we need to check for remaining 2 bytes
+
+            if (localBuffer.Length < 2)
+            {
+                return;
+            }
+
+            byte next = localBuffer[1];
+            if (localBuffer[0] == 0x80 && (next == 0xA8 || next == 0xA9))
```
Comment (krwq) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs]: (pending braces around the throw)

**Comment by krwq** [comment]
yes, this can wait, since there is currently no perf tests related to comments I'll let you know when I write them 😄

**Comment by krwq** [comment]
Perf results are here: https://github.com/dotnet/performance/pull/543
Multi-segment multiline comment specifically seems to be 2x faster for the common scenario, everything else is mostly the same. For some reason single-segment multiline comment regressed a bit (I have not changed that code path or anything related and have tried re-running tests)

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs (hunk 1 above)
Comment (krwq) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: initially changed to that but later have removed ConsumeEmptySingleLineComment completely - once I unified consume and skip it was almost no op (only if-statement was there)

**Comment by ahsonkhan** [comment]
> Multi-segment multiline comment specifically seems to be 2x faster for the common scenario, everything else is mostly the same. For some reason single-segment multiline comment regressed a bit (I have not changed that code path or anything related and have tried re-running tests)

I ran the perf tests locally and was not able to reproduce your results:

summary:
better: 1, geomean: 1.032
worse: 9, geomean: 1.073
total diff: 10

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Tests.Utf8JsonReaderCommentsTests.Utf8JsonReaderCommentParsing( |      1.15 |           169.56 |           194.15 |         |
| System.Text.Json.Tests.Utf8JsonReaderCommentsTests.Utf8JsonReaderCommentParsing( |      1.14 |           920.21 |          1048.72 |         |
| System.Text.Json.Tests.Utf8JsonReaderCommentsTests.Utf8JsonReaderCommentParsing( |      1.08 |           118.35 |           127.87 |         |
| System.Text.Json.Tests.Utf8JsonReaderCommentsTests.Utf8JsonReaderCommentParsing( |      1.07 |           143.18 |           152.90 |         |
| System.Text.Json.Tests.Utf8JsonReaderCommentsTests.Utf8JsonReaderCommentParsing( |      1.07 |           172.37 |           183.75 |         |
| System.Text.Json.Tests.Utf8JsonReaderCommentsTests.Utf8JsonReaderCommentParsing( |      1.05 |           150.26 |           157.17 |         |
| System.Text.Json.Tests.Utf8JsonReaderCommentsTests.Utf8JsonReaderCommentParsing( |      1.04 |           142.63 |           148.78 |         |
| System.Text.Json.Tests.Utf8JsonReaderCommentsTests.Utf8JsonReaderCommentParsing( |      1.04 |           147.42 |           153.37 |         |
| System.Text.Json.Tests.Utf8JsonReaderCommentsTests.Utf8JsonReaderCommentParsing( |      1.04 |           121.93 |           126.25 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Tests.Utf8JsonReaderCommentsTests.Utf8JsonReaderCommentParsing( |      1.03 |           131.92 |           127.82 |         |

**Comment by adamsitnik** [comment]
>  it looks like we are hitting a character limit on the ResultComparer output

I am not sure what would be the best way to solve it. The default output is console, which has a limited width. Most probably the easiest thing to do right now would be to shorten the ID.

```log
System.Text.Json.Tests.Utf8JsonReaderCommentsTests.Utf8JsonReaderCommentParsing
System.Text.Json.Tests.CommentsTests.Parsing
```

> can the ResultComparer store the result to disk somewhere

yes, you can pass the path to output csv file via `--csv $filePath` ([full docs](https://github.com/dotnet/performance/tree/master/src/tools/ResultsComparer))

**Comment by ahsonkhan** [comment]
> yes, you can pass the path to output csv file via `--csv $filePath`

> (by default - maybe md and csv) just like the actual perf runs (rather than us having to specify a path - maybe in one of the base/diff paths)

I am aware of the `--csv` option. I was asking, by default, without having to pass that option in.

**Comment by krwq** [comment]
@ahsonkhan ping, is there anything blocking here? I'm not personally concerned about tiny (< 20%) regression in perf here considering comments are not expected to be used in high perf scenarios and there have been couple of reliability fixes here

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 12]
```diff
@@ -2166,195 +2166,53 @@ private ConsumeTokenResult ConsumeNextTokenUntilAfterAllCommentsAreSkippedMultiS
...
 
-            if (marker != JsonConstants.Asterisk)
-            {
-                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedStartOfValueNotFound, JsonConstants.Slash);
-            }
-
-            return SkipMultiLineCommentMultiSegment(localBuffer.Slice(1), leftOver);
-        }
+                ReadOnlySequence<byte> commentSequence = _sequence.Slice(start, end);
+                commentSequence = commentSequence.Slice(2, commentSequence.Length - 2 - tailBytesToIgnore);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: Can we try to avoid the double slicing of the `ReadOnlySequence`?
Maybe we can already have the start `SequencePosition` be off-set by 2. and change `end` to account for the `- 2 - tailBytesToIgnore`

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 13]
```diff
@@ -2166,195 +2166,53 @@ private ConsumeTokenResult ConsumeNextTokenUntilAfterAllCommentsAreSkippedMultiS
...
-                        return false;
-                    }
-                    _totalConsumed += localBuffer.Length + leftOver;
-                    _bytePositionInLine += localBuffer.Length + leftOver;
-                    leftOver = 0;
-                    localBuffer = _buffer;
-                }
+                _totalConsumed = prevTotalConsumed;
+                // consumed might not be correct now but it won't be used anymore after we return false
+                // if we don't wrong number of bytes might get reported (if consuming called GetNextSpan)
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: I have a hard time understanding what this comment is saying.
`If we don't... wrong number`, if we don't do what?
And what do you mean by `consuming` in `if consuming called GetNextSpan`

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 14]
```diff
@@ -2449,207 +2332,242 @@ private bool ConsumeSingleLineCommentMultiSegment(ReadOnlySpan<byte> localBuffer
...
+            }
+
+            int totalIdx = 0;
+            while (true)
+            {
+                int idx = localBuffer.IndexOfAny(JsonConstants.LineFeed, JsonConstants.CarriageReturn, JsonConstants.StartingByteOfNonStandardSeparator);
+                dangerousLineSeparatorBytesConsumed = 0;
+
+                if (idx == -1)
+                    return -1;
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: nit: add braces here/elsewhere (whenever you have an if within a nested scope, even if it happens to be a single line).

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 15]
```diff
@@ -2449,207 +2332,242 @@ private bool ConsumeSingleLineCommentMultiSegment(ReadOnlySpan<byte> localBuffer
...
+
+                if (localBuffer[idx] != JsonConstants.StartingByteOfNonStandardSeparator)
+                    return totalIdx + idx;
+
+                int p = idx + 1;
+                localBuffer = localBuffer.Slice(p);
+                totalIdx += p;
+
+                // or dangerousLineSeparatorBytesConsumed = 1
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: What do you mean by this comment?

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 16]
```diff
@@ -2449,207 +2332,242 @@ private bool ConsumeSingleLineCommentMultiSegment(ReadOnlySpan<byte> localBuffer
...
+                {
+                    // no match
+                    dangerousLineSeparatorBytesConsumed = 0;
+                    return;
+                }
+            }
+
+            if (dangerousLineSeparatorBytesConsumed == 2)
+            {
+                if (localBuffer[0] == 0xA8 || localBuffer[0] == 0xA9)
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: nit: store `localBuffer[0]` in a local to avoid double access.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 17]
```diff
@@ -2449,207 +2332,242 @@ private bool ConsumeSingleLineCommentMultiSegment(ReadOnlySpan<byte> localBuffer
...
+                        case JsonConstants.Asterisk:
+                            expectSlash = true;
+                            _bytePositionInLine += nextIdx;
+                            break;
+                        case JsonConstants.LineFeed:
+                            _bytePositionInLine = 0;
+                            _lineNumber++;
+                            break;
+                        default: // JsonConstants.CarriageReturn:
+                            _bytePositionInLine = 0;
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: nit: add `Debug.Assert(marker == JsonConstants.CarriageReturn)` and remove the comment.

File: src/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs [hunk 18]
```diff
@@ -884,6 +885,262 @@ public static void SkipSingleLineCommentMultiSpanTest(string lineSeparator)
...
+        public static void JsonWithSingleLineCommentInTheMiddleOfThePayloadEndingWithNonStandardLineEndingMultiSegment(string jsonString, int segmentSize)
+        {
+            byte[] dataUtf8 = Encoding.UTF8.GetBytes(jsonString);
+            ReadOnlySequence<byte> sequence = JsonTestHelper.GetSequence(dataUtf8, segmentSize);
+
+            foreach (JsonCommentHandling jsonCommentHandling in typeof(JsonCommentHandling).GetEnumValues())
+            {
+                var state = new JsonReaderState(options: new JsonReaderOptions { CommentHandling = jsonCommentHandling });
+
+
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs]: nit: remove extra new line

File: src/System.Text.Json/tests/Utf8JsonReaderTests.cs [hunk 19]
```diff
@@ -3004,7 +3086,8 @@ public static void ConsumeSingleLineCommentSingleSpanTest(string lineSeparator)
                 var json = new Utf8JsonReader(dataUtf8.AsSpan(0, i), isFinalBlock: false, state);
                 VerifyReadLoop(ref json, expected);
 
-                json = new Utf8JsonReader(dataUtf8.AsSpan((int)state.BytesConsumed), isFinalBlock: true, state);
+                Span<byte> remainder = dataUtf8.AsSpan((int)state.BytesConsumed);
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Utf8JsonReaderTests.cs]: nit: `ReadOnlySpan<byte>`

**Comment by ahsonkhan** [comment]
> is there anything blocking here? 

Nope, nothing blocking.

> I'm not personally concerned about tiny (< 20%) regression in perf here considering comments are not expected to be used in high perf scenarios and there have been couple of reliability fixes here

Agreed. I am still a bit concerned about the regression (especially for the "skip" mode), but we can work on perf optimizations separately.

Do you have the latest numbers based on your changes? We should double check if we have 10-20% regression on the "skip" case, because that is a bit too high.

**Comment by ahsonkhan** [comment]
@krwq, any updates to this?

**Comment by krwq** [comment]
@ahsonkhan, I have not had a chance to re-run this yet. Nothing has changes since you've run it yourself last time

**Comment by ahsonkhan** [comment]
> I have not had a chance to re-run this yet. Nothing has changes since you've run it yourself last time

OK. Let me know when you have addressed the leftover feedback so we can merge this change in :)
Thanks.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs (hunk 12 above)
Comment (krwq) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: I don't believe this is possible. SequencePosition doesn't know how to move by 2 since it may end up in different object which it doesn't know about (where ReadOnlySequence does)

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs (hunk 15 above)
Comment (krwq) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: removing since it's not super valuable and I can't remember myself

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs (hunk 13 above)
Comment (krwq) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: yep, re-worded with more details

**Comment by krwq** [comment]
@ahsonkhan seems the conflict is that we repeated this fix:
https://github.com/dotnet/corefx/pull/38036/files#diff-b12b49879ec5ef146f3373b69e66216bL3007

I'm building locally to double check my tests still build and run correctly after rebasing.
